esp/esp/program/controllers/autoscheduler/scoring.py

```diff
@@ -0,0 +1,1136 @@
+        return self.total_score / self.num_sections
+
+    def process_section(self, section, room):
+        """Computes the score for a given section assigned to the given
+        room."""
+        # Return the fraction of satisfied resource requests.
+        met_criteria = 0.0
+        requested_criteria = 0.0
+        for request in section.resource_requests.itervalues():
+            if request.name not in self.requests_to_ignore \
+                    and request.value != "":
+                requested_criteria += 1
+                if request.name in room.furnishings \
+                        and room.furnishings[request.name].value \
+                        == request.value:
+                    met_criteria += 1
+        if requested_criteria == 0.0:
+            return 1.0
+        else:
+            return met_criteria / requested_criteria
+
+    @util.timed_func("ResourceValueMatchingScorer")
+    def update_schedule(self, schedule):
+        """Overwrite internal state to reflect the given schedule."""
+        # Scaling is just the default 1 because we're just scoring by section.
+        self.num_sections = len(schedule.class_sections)
+
+        furnishing_values = {}
+        for room in schedule.classrooms.itervalues():
+            for furnishing in room.furnishings.itervalues():
+                if furnishing.name not in furnishing_values:
+                    furnishing_values[furnishing.name] = set()
+                furnishing_values[furnishing.name].add(furnishing.value)
+        self.requests_to_ignore = set()
+        for name, values in furnishing_values.iteritems():
+            if len(values) == 1:
+                self.requests_to_ignore.add(name)
+        self.total_score = 0.0
+        for section in schedule.class_sections.itervalues():
+            if section.is_scheduled():
+                self.total_score += self.process_section(
+                    section, section.assigned_roomslots[0].room)
+
+    @util.timed_func("ResourceValueMatchingScorer")
+    def update_schedule_section(self, section, start_roomslot):
+        """Update the internal state to reflect the scheduling of the specified
+        section to start at the specified roomslot."""
+        self.total_score += self.process_section(section, start_roomslot.room)
+
+    @util.timed_func("ResourceValueMatchingScorer")
+    def update_unschedule_section(self, section):
+        """Update the internal state to reflect the uncsheduling of the
+        specified section."""
+        self.total_score -= self.process_section(
+            section, section.assigned_roomslots[0].room)
+
+
+class RoomConsecutivityScorer(BaseScorer):
+    """Try to schedule classes consecutively in classrooms. This is a good
+    heuristic for being able to schedule long classes later, as well as for
+    minimizing having to repeatedly clean or lock/unlock rooms."""
+
+    def boundary_complete_before(self, roomslot, ignore_section=None):
```

Comment:
can you define what it means for a boundary to be "complete" sooner in the code?

Reply:
Better?